### PR TITLE
Add proxy for BYOS instances

### PR DESCRIPTION
**Pull request by jesusbv**
- Add engine to handle proxy byos systems
- Register system with SCC if proxy byos system
- Activate the product
- Update subscription from SCC
- Add tests

## Description

When handling proxy BYOS instances, it would be a good idea to track those in the database
as they follow different paths than regular instances, also there is new registration and 
validation of those instances through SCC

## Change Type

*Please select the correct option.*

- [ ] **Bug Fix** (a non-breaking change which fixes an issue)
- [X] **New Feature** (a non-breaking change which adds new functionality)
- [ ] **Documentation Update** (a change which only updates documentation)

## Checklist

*Please check off each item if the requirement is met.*

- [X] I have verified that my code follows RMT's coding standards with `rubocop`.
- [X] I have reviewed my own code and believe that it's ready for an external review.
- [X] I have provided comments for any hard-to-understand code.
- [ ] I have documented the `MANUAL.md` file with any changes to the user experience.
- [X] RMT's test coverage remains at 100%.
- [ ] If my changes are non-trivial, I have added a changelog entry to notify users at `package/obs/rmt-server.changes`.

## Other Notes

Happy to update changelog if decided so

### CODE REVIEW

File: engines/instance_verification/lib/instance_verification/engine.rb [hunk 1]
```diff
@@ -1,4 +1,17 @@
+require 'net/http'
+
+ACTIVATE_PRODUCT_URL = 'https://scc.suse.com/connect/systems/products'.freeze
+SYSTEM_SUBSCRIPTION_URL = 'https://scc.suse.com/connect/systems/subscriptions'.freeze
+SUBSCRIPTIONS_PRODUCTS_URL = 'https://scc.suse.com/connect/subscriptions/products'.freeze
```
Comment (digitaltom) [inline engines/instance_verification/lib/instance_verification/engine.rb:5]: Why does instance_verification need to access SCC? I thought it's checking with the instance data, if the client system is allowed to use a specific product?

File: engines/instance_verification/lib/instance_verification/engine.rb [hunk 2]
```diff
@@ -39,6 +52,7 @@ def verify_product_activation
             verify_extension_activation(product)
           end
         rescue InstanceVerification::Exception => e
+          raise ActionController::TranslatedError.new(e.message) if @system.proxy_byos
```
Comment (digitaltom) [inline engines/instance_verification/lib/instance_verification/engine.rb:55]: Is a BYOS system actually expected to pass the instance verification or not? 
If not, we could bypass the whole instance verification in that case and also don't need to send `--instance-data` from SUSEConnect. (@rjschwei ?)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/219815/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-run-tests. (Merged 042ca33fdbf68b5610f285e469fbaa7ac1295c31 into 36cce88c8cbddcdbcde9051b9dea98095183be53)

File: engines/instance_verification/lib/instance_verification/engine.rb [hunk 3]
```diff
@@ -49,6 +63,44 @@ def verify_product_activation
...
+          http.use_ssl = true
+          scc_request = prepare_scc_request(uri.path, product)
+          response = http.request(scc_request)
+          logger.info(
+            "Response code is #{response.code} for the attempt to " \
+              "check registration code for #{product.product_string} to SCC"
+          )
+          raise InstanceVerification::Exception.new('No subscription with this Registration Code found') unless response.code_type == Net::HTTPOK
+        end
+
```
Comment (digitaltom) [inline engines/instance_verification/lib/instance_verification/engine.rb:103]: Can those methods also move to the `scc_proxy` engine? They seem unrelated to instance verification.

File: engines/scc_proxy/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# SccProxy
+Short description and motivation.
```
[the remark this answers is not in the document]
Comment (digitaltom) [reply engines/scc_proxy/README.md:2]: Yes, please describe the use case and maybe link to the confluence page with further details. `byos_proxy` might be a more descriptive name.

File: engines/scc_proxy/bin/rails [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+ENGINE_ROOT = File.expand_path('..', __dir__)
+ENGINE_PATH = File.expand_path('../lib/scc_proxy/engine', __dir__)
+
+# Set up gems listed in the Gemfile.
+ENV['BUNDLE_GEMFILE'] ||= File.expand_path('../Gemfile', __dir__)
+require 'bundler/setup' if File.exist?(ENV['BUNDLE_GEMFILE'])
+
+require 'rails'
+require 'rails/engine/commands'
```
Comment (digitaltom) [inline engines/scc_proxy/bin/rails:13]: Do we need this file?

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 6]
```diff
@@ -0,0 +1,140 @@
+require 'net/http'
+
+REGISTER_URL = 'https://scc.suse.com/connect/subscriptions/systems'.freeze
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:3]: Better: `ANNOUNCE_URL`

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+    def headers(auth)
+      {
+        'accept' => 'application/json,application/vnd.scc.suse.com.v4+json',
+        'Content-Type' => 'application/json',
+        'Authorization' => auth
+      }
+    end
+
+    def prepare_scc_registration_request(uri_path, auth, params)
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:34]: Also here, better to call it `prepare_scc_announce_request`.

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        identifier: product.identifier,
+        version: product.version,
+        arch: product.arch,
+        release_type: product.release_type,
+        email: email || nil
+      }.to_json
+      scc_request
+    end
+
+    def register_system_scc(auth, params)
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:58]: Better: `announce_system_scc`.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/219816/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 78479e15929d9b8b322c346e1cfd2257e5dcb2f9 into 36cce88c8cbddcdbcde9051b9dea98095183be53)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/219817/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 65cfc8a85804b8524defc5bc2270dd4830474dd5 into 36cce88c8cbddcdbcde9051b9dea98095183be53)

File: engines/instance_verification/lib/instance_verification/engine.rb (hunk 2 above)
Comment (rjschwei) [reply engines/instance_verification/lib/instance_verification/engine.rb:55]: @digitaltom A BYOS instance is not expected to pass the instance verification. However we have the case where a user may pass a regcode while running on a PAYG instance. As such before doing regcode validation we have to first ensure that PAYG instance verification fails. Otherwise we may put a user into a situation where they end up having an instance counted against their subscription while also paying through the CSP.

File: engines/instance_verification/lib/instance_verification/engine.rb (hunk 3 above)
Comment (jesusbv) [reply engines/instance_verification/lib/instance_verification/engine.rb:103]: Yes, that is the idea, do a refactor after merging this PR

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/220273/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-run-tests. (Merged d76598928a5151b22e780a5b19208b3fce02be44 into 36cce88c8cbddcdbcde9051b9dea98095183be53)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/220292/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged d76598928a5151b22e780a5b19208b3fce02be44 into 36cce88c8cbddcdbcde9051b9dea98095183be53)

File: engines/instance_verification/lib/instance_verification/engine.rb (hunk 2 above)
Comment (digitaltom) [reply engines/instance_verification/lib/instance_verification/engine.rb]: Could it be that the customer intentionally uses the registration code on PAYG, for example to receive priority support? 
I suggest to clearly document in the code that we do instance verification first and try BYOS with the regcode afterwards because of the use case you described.

Comment (rjschwei) [reply engines/instance_verification/lib/instance_verification/engine.rb]: +1 on doc via comments.

Customers that use PAYG get L1 & L2 support through the CSP. If they want direct support by SUSE for those level they need to switch to BYOS.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/220309/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 09a6950bf8f34eb96bab7022a78228031cac3c43 into 36cce88c8cbddcdbcde9051b9dea98095183be53)

File: engines/instance_verification/lib/instance_verification/engine.rb [hunk 9]
```diff
@@ -39,7 +50,17 @@ def verify_product_activation
             verify_extension_activation(product)
           end
         rescue InstanceVerification::Exception => e
-          raise ActionController::TranslatedError.new('Instance verification failed: %{message}' % { message: e.message })
+          if @system.proxy_byos
+            # BYOS instances that use RMT as a proxy are expected to fail the
+            # instance verification check, however, PAYG instances may send registration
+            # code, as such, instance verification engine checks for those BYOS
+            # instances once instance verification has failed
+            raise ActionController::TranslatedError.new('No subscription with this Registration Code found') unless scc_check_regcode(product)
```
Comment (digitaltom) [inline engines/instance_verification/lib/instance_verification/engine.rb:58]: Why do we need to do the `scc_check_regcode` check here at all? 
If the regcode is invalid, or doesn't include the requested product then `SccProxy.scc_activate_product` will fail accordingly.

File: engines/instance_verification/lib/instance_verification/engine.rb [hunk 10]
```diff
@@ -1,4 +1,15 @@
+require 'net/http'
+
+SUBSCRIPTIONS_PRODUCTS_URL = 'https://scc.suse.com/connect/subscriptions/products'.freeze
+
 module InstanceVerification
+  extend ::Net::HTTPHeader
+
+  def self.verification_basic_encode(login, password)
```
Comment (digitaltom) [inline engines/instance_verification/lib/instance_verification/engine.rb:8]: Is this method used from anywhere? I wasn't able to find a usage.

File: engines/instance_verification/spec/requests/api/connect/v3/subscriptions/systems_controller_spec.rb [hunk 11]
```diff
@@ -4,11 +4,94 @@
...
+    context 'using RMT generated credentials' do
+      it 'saves instance data' do
+        post '/connect/subscriptions/systems', params: { hostname: 'test', instance_data: instance_data }
+        data = JSON.parse(response.body)
+        system = System.find_by(login: data['login'])
+        expect(system.hw_info.instance_data).to eq(instance_data)
+      end
+    end
+
+    context 'registering the system with SCC' do
```
Comment (digitaltom) [inline engines/instance_verification/spec/requests/api/connect/v3/subscriptions/systems_controller_spec.rb:16]:
```suggestion
    context 'using SCC generated credentials (BYOS mode)' do
```

Comment (digitaltom) [reply engines/instance_verification/spec/requests/api/connect/v3/subscriptions/systems_controller_spec.rb:16]: I think this test needs to move to the scc_proxy spec. 
In the instance_verification spec, we would only need to make sure that a PAYG system that provides a regcode doesn't fail.

File: engines/instance_verification/spec/requests/api/connect/v3/systems/products_controller_spec.rb [hunk 12]
```diff
@@ -108,6 +128,98 @@
           expect(response.body).to eq(serialized_service_sap_json)
         end
       end
+
+      context 'when system is connected to SCC' do
```
Comment (digitaltom) [inline engines/instance_verification/spec/requests/api/connect/v3/systems/products_controller_spec.rb:132]: This also needs to move to the `scc_proxy` spec I think, because all the code it tests is from there.

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+      uri = URI.parse(ACTIVATE_PRODUCT_URL)
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      scc_request = prepare_scc_request(uri.path, product, auth, token, email)
+      response = http.request(scc_request)
+      logger.info(
+        "Response code is #{response.code} for the attempt to " \
+          "activate product #{product.product_string} to SCC"
+      )
+      return false unless response.code_type == Net::HTTPCreated
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:80]: I think it would be helpful to add detailed logging of the response in case the result is not successful.

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+    isolate_namespace SccProxy
+    config.generators.api_only = true
+
+    config.after_initialize do
+      # replaces RMT registration for SCC registration
+      Api::Connect::V3::Subscriptions::SystemsController.class_eval do
+        def announce_system
+          auth_header = request.headers['HTTP_AUTHORIZATION'] if request.headers.include?('HTTP_AUTHORIZATION')
+          auth_header ||= '='
+          auth_header = auth_header[(auth_header.index('=') + 1)..-1]
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:97]: What is this doing? Maybe create a helper method for it with a descriptive name.

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+          if auth_header.empty?
+            # no token sent to check with SCC
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:99]: In which use case do we end up here, creating the system without announcing to SCC?

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        rescue *NET_HTTP_ERRORS => e
+          logger.error(
+            "Could not register system with regcode #{auth_header} " \
+              "to SCC: #{e.message}"
+          )
+          if e.message.include? '401'
+            # same message and status than SUSEConnect cli output
+            # check why it is 422 and not 401
+            status = 422
+            message = 'No subscription with this Registration Code found'
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:122]: Better copy the error message from the server, because it could also be the the registration code expired, is not yet activated etc.

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+          else
+            message = e.message
+            status = e.message[0..(message.index(' ') - 1)].to_i
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:125]: Why do we need to parse the status from the message, can't we just use the status from the original response?

File: engines/instance_verification/lib/instance_verification/engine.rb (hunk 10 above)
Comment (jesusbv) [reply engines/instance_verification/lib/instance_verification/engine.rb:8]: Yes, forgot to remove during the cleaning of the engine

File: engines/scc_proxy/lib/scc_proxy/engine.rb (hunk 15 above)
Comment (jesusbv) [reply engines/scc_proxy/lib/scc_proxy/engine.rb:99]: for PAYG instances ? those still need to register normally, right ?
for PAYG instances that header is `Token token=` so if no regcode, should do the same as [regular path](https://github.com/SUSE/rmt/blob/master/app/controllers/api/connect/v3/subscriptions/systems_controller.rb#L3)

File: engines/scc_proxy/lib/scc_proxy/engine.rb (hunk 14 above)
Comment (digitaltom) [reply engines/scc_proxy/lib/scc_proxy/engine.rb:97]: Ah, it would be easier to understand if you wrap it in a method like `byos_request?`

File: engines/scc_proxy/lib/scc_proxy/engine.rb (hunk 15 above)
Comment (digitaltom) [reply engines/scc_proxy/lib/scc_proxy/engine.rb:99]: Ah, I see. Could you please add a comment that this is a copy of the standard announce code? Is there a way not to duplicate it but run through the normal announce code in that case?

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/220587/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged a4110aef2312aaf7b7b3c6de0d530ed2485f8ebd into 36cce88c8cbddcdbcde9051b9dea98095183be53)

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+          render json: { type: 'error', error: e.message }, status: status_code(e.message), location: nil
+        end
+
+        protected
+
+        def status_code(error_message)
+          error_message[0..(error_message.index(' ') - 1)].to_i
+        end
+
+        def header_has_regcode?(auth_header)
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:118]: I think that name is wrong, it checks for `has_no_regcode`.

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+          if @system.proxy_byos
+            response = SccProxy.scc_activate_product(@product, auth, params[:token], params[:email])
+            unless response.code_type == Net::HTTPCreated
+              error = JSON.parse(response.body)
+              logger.info "Could not activate #{@product.product_string}, error: #{error['error']} #{response.code}"
+              error['error'] = parse_error(error['error']) if error['error'].include? 'json'
+              raise ActionController::TranslatedError.new(error['error'])
+            end
+            logger.info "Product #{@product.product_string} successfully activated with SCC"
+            InstanceVerification.update_cache(request.remote_ip, @system.login, @product.id)
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:143]: Why is this needed?

**Comment by digitaltom** [comment]
Looks ok now conceptually. I think the new `engines/scc_proxy` needs a proper README that describes the use cases and functionality of the engine.
What I'm lacking is a way to fully test this. It would be helpful to have a section in the README that describes a local dev setup, for example like: 

* Start RMT with enabled engines: `RMT_LOAD_ENGINES=true bin/rails s -b 0.0.0.0`
* Mirror SLE15SP2
* Register a BYOS client system with it: 
```
docker run --rm --privileged --network=host -ti registry.scc.suse.de/suse/sles15:15.2
SUSEConnect --url=http://172.17.0.1:4224 -r <sles_regcode>
``` 
 
Currently this fails with: `Unknown error reading from 'plugin:/susecloud?credentials=SUSE_Linux_Enterprise_Server_x86_64&path=/services/1939'` 
Do you know how to get that susecloud zypper plugin working? Why is that even needed? 

Before merging I suggest you run through a full test in your cloud setup: 

* Register BYOS system with invalid regcode and see it failing
* Register with correct regcode, see it succeed, check the system in the SCC webui
* Try to activate a paid extension without + with invalid regcode, see it fail
* Activate a paid extension (eg. HA) with corect regcode, see it succeed, check it in SCC webui

**Comment by jesusbv** [comment]
> Looks ok now conceptually. I think the new `engines/scc_proxy` needs a proper README that describes the use cases and functionality of the engine. What I'm lacking is a way to fully test this. It would be helpful to have a section in the README that describes a local dev setup, for example like:
> 
>     * Start RMT with enabled engines: `RMT_LOAD_ENGINES=true bin/rails s -b 0.0.0.0`
> 
>     * Mirror SLE15SP2
> 
>     * Register a BYOS client system with it:
> 
> 
> ```
> docker run --rm --privileged --network=host -ti registry.scc.suse.de/suse/sles15:15.2
> SUSEConnect --url=http://172.17.0.1:4224 -r <sles_regcode>
> ```
> 
> Currently this fails with: `Unknown error reading from 'plugin:/susecloud?credentials=SUSE_Linux_Enterprise_Server_x86_64&path=/services/1939'` Do you know how to get that susecloud zypper plugin working? Why is that even needed?

Yes, you need [URL resolver plugin](https://github.com/SUSE-Enceladus/cloud-regionsrv-client/blob/master/usr/lib/zypp/plugins/urlresolver/susecloud) included in [cloud-regionsrv-client](https://github.com/SUSE-Enceladus/cloud-regionsrv-client) to translate `plugin:/susecloud` to a proper url

> 
> Before merging I suggest you run through a full test in your cloud setup:
> 
* Register BYOS system with invalid regcode and see it failing  :heavy_check_mark: 

* Register with correct regcode, see it succeed, :heavy_check_mark: 
*  check the system in the SCC webui :x:  (not yet) 

>     * Try to activate a paid extension without + with invalid regcode, see it fail
> 
>     * Activate a paid extension (eg. HA) with corect regcode, see it succeed, check it in SCC webui
Not yet these

File: engines/scc_proxy/lib/scc_proxy/engine.rb (hunk 19 above)
Comment (jesusbv) [reply engines/scc_proxy/lib/scc_proxy/engine.rb:143]: I would say so, otherwise, `ZypperAuth` would try to verify the instance _again_ ([check here](https://github.com/SUSE/rmt/blob/master/engines/zypper_auth/lib/zypper_auth/engine.rb#L17))
and we would need to do the whole process again

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/220673/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 8412aea48dcafc15135cb3b7e1733aa6ff92ed15 into 36cce88c8cbddcdbcde9051b9dea98095183be53)

File: engines/scc_proxy/lib/scc_proxy/engine.rb (hunk 19 above)
Comment (digitaltom) [reply engines/scc_proxy/lib/scc_proxy/engine.rb:143]: What would happen after the 20 minute expiry?

**Comment by rjschwei** [comment]
> 
> * Start RMT with enabled engines: `RMT_LOAD_ENGINES=true bin/rails s -b 0.0.0.0`
> * Mirror SLE15SP2
> * Register a BYOS client system with it:
> 
> ```
> docker run --rm --privileged --network=host -ti registry.scc.suse.de/suse/sles15:15.2
> SUSEConnect --url=http://172.17.0.1:4224 -r <sles_regcode>
> ```
> 
> Currently this fails with: `Unknown error reading from 'plugin:/susecloud?credentials=SUSE_Linux_Enterprise_Server_x86_64&path=/services/1939'` Do you know how to get that susecloud zypper plugin working? Why is that even needed?

For the test case a simple url resolver plugin could be written and in the test case it would simply add the url.

In production the plugin handles failover. If the target update server is down for some reason the plugin will attempt to find another update server.

**Comment by digitaltom** [comment]
> For the test case a simple url resolver plugin could be written and in the test case it would simply add the url.
> 
> In production the plugin handles failover. If the target update server is down for some reason the plugin will attempt to find another update server.

Is there any trick to get that plugin working locally with a local RMT? It would speed up development quite a lot if it would be possible to run + debug locally.

**Review by digitaltom — APPROVED** [review]
Please check what happens after the 20m cache timeout and add a description to the README.

**Comment by rjschwei** [comment]
> > For the test case a simple url resolver plugin could be written and in the test case it would simply add the url.
> > In production the plugin handles failover. If the target update server is down for some reason the plugin will attempt to find another update server.
> 
> Is there any trick to get that plugin working locally with a local RMT? It would speed up development quite a lot if it would be possible to run + debug locally.

No, the production implementation of the plugin does not have a dev shortcut setting. I think it would be best to implement a plugin as part of the test suite in RMT.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/220947/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 8412aea48dcafc15135cb3b7e1733aa6ff92ed15 into 4a25d1d5063caf8af7469d752e81085e530da82a)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/220978/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged a9181ab34bfa12bda5ac70b0d3afad89065413dc into 4a25d1d5063caf8af7469d752e81085e530da82a)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/220979/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged e484ccb25607dd4e49d4d6ac8840ed9b8b93e6a7 into 4a25d1d5063caf8af7469d752e81085e530da82a)

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 20]
```diff
@@ -74,6 +75,32 @@ def scc_activate_product(product, auth, token, email)
...
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      scc_request = Net::HTTP::Get.new(uri.path, headers(auth))
+      response = http.request(scc_request)
+      unless response.code_type == Net::HTTPOK
+        logger.info "Could not get the system (#{login}) subscription, error: #{response.message} #{response.code}"
+        response.message = SccProxy.parse_error(response.message) if response.message.include? 'json'
+        return { is_active: false, message: response.message }
+      end
+      parsed_response = JSON.parse(response.body)[0]
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:98]: You're always checking the first returned subscription here, that won't work in case the system is using multiple subscriptions. For example when the system uses one subscription for SLES and another subscription for and extension like HA.

File: engines/zypper_auth/lib/zypper_auth/engine.rb [hunk 21]
```diff
@@ -27,12 +27,21 @@ def verify_instance(request, logger, system)
       Rails.cache.write(cache_key, is_valid, expires_in: 20.minutes)
       is_valid
     rescue InstanceVerification::Exception => e
+      message = ''
+      if system.proxy_byos
+        result = SccProxy.scc_check_subscription_expiration(request.headers, system.login, logger) if system.proxy_byos
+        return true if result[:is_active]
```
Comment (digitaltom) [inline engines/zypper_auth/lib/zypper_auth/engine.rb:33]: This is not getting put in the cache, so would the check get performed on each zypper access to the repository? That can get pretty slow.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/221582/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged ee8a700d4b7d29190f816178c1306eceff8bcbd8 into 4a25d1d5063caf8af7469d752e81085e530da82a)

File: engines/zypper_auth/lib/zypper_auth/engine.rb (hunk 21 above)
Comment (jesusbv) [reply engines/zypper_auth/lib/zypper_auth/engine.rb]: In case the subscription is active, cache gets updated, other than that, yes system subscriptions needs to get checked

File: engines/scc_proxy/lib/scc_proxy/engine.rb (hunk 20 above)
Comment (jesusbv) [reply engines/scc_proxy/lib/scc_proxy/engine.rb]: Fixed

**Comment by jesusbv** [comment]
> Do we also have official customer docs that we can link here?

Not yet

> Did you try this out? After SUSEConnect -d (--de-register) zypper will not have any repos available anymore to install that package.

Yes, I have, I copied the wrong line. Fixed.

> Is registercloudguest available on all cloud AMIs, or when do we have to install it?

The package is installed in the Public Cloud PAYG images, if BYOS the package should be installed, long term the package will also be installed in BYOS (no further action then)

> -d -p will de-register a product.

Yes. I don't know what you want to say with that, sorry. If it is regarding the README commands, I just out some examples of how to use

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/221586/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged b6fa2e73cd2bfbab1e74424d7f3d099f435dcc87 into 4a25d1d5063caf8af7469d752e81085e530da82a)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/221587/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 03fa6dc3b4b0fe9722e348d4e8d7392b526e0779 into 4a25d1d5063caf8af7469d752e81085e530da82a)

File: engines/instance_verification/lib/instance_verification/engine.rb [hunk 22]
```diff
@@ -1,4 +1,10 @@
 module InstanceVerification
+  def self.update_cache(remote_ip, system_login, product_id)
+    cache_key = [remote_ip, system_login, product_id].join('-')
+    # caches verification result to be used by zypper auth plugin
+    Rails.cache.write(cache_key, true, expires_in: 20.minutes)
```
Comment (digitaltom) [inline engines/instance_verification/lib/instance_verification/engine.rb:5]: Is there a reason behind having this cache for 20 minutes and not for 24h for example?

Comment (jesusbv) [reply engines/instance_verification/lib/instance_verification/engine.rb:5]: It was one hour before (2 years ago) and it's been 20 min for that long.
Although, that change would be out of scope for this PR, it would need its own issue/PR

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/221735/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged cc99b0e6393b4cc771a915a0516953185569bd72 into 4a25d1d5063caf8af7469d752e81085e530da82a)

File: engines/instance_verification/lib/instance_verification/engine.rb (hunk 22 above)
Comment (rjschwei) [reply engines/instance_verification/lib/instance_verification/engine.rb:5]: The 20 minutes is a more or less arbitrary value. We expect that one run of `zypper` completes in this time meaning we do the system lookup only once. Also it reduces the window where the credentials and metadata can be injected into another system.

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+      response = http.request(scc_request)
+      unless response.code_type == Net::HTTPOK
+        logger.info "Could not get the system (#{login}) subscription, error: #{response.message} #{response.code}"
+        response.message = SccProxy.parse_error(response.message) if response.message.include? 'json'
+        return { is_active: false, message: response.message }
+      end
+      array_subscriptions = JSON.parse(response.body)
+      array_subscriptions.each do |subscription|
+        match = subscription['systems'].each do |system_subscribed|
+          true if system_subscribed['login'] == login
```
[the remark this answers is not in the document]
Comment (jesusbv) [reply engines/scc_proxy/lib/scc_proxy/engine.rb:101]: the check for the subscription to be valid (active, or not expired) is [there](https://github.com/SUSE/rmt/blob/handle-proxy-byos/engines/zypper_auth/lib/zypper_auth/engine.rb#L33) ?
The line 101, is to make sure we check the active/expire of the current subscription being used in the request.

WRT the repository, the check is in place to make sure the system has an active subscription

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/221747/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged cc99b0e6393b4cc771a915a0516953185569bd72 into e30273fcb614dc9ea2561a4d2a903122f5669f3b)

File: engines/instance_verification/lib/instance_verification/engine.rb (hunk 22 above)
Comment (bear454) [reply engines/instance_verification/lib/instance_verification/engine.rb:5]: Please be aware that the cache expiration is not respected when using the default file system cache, and if we want it changed, it needs to be changed in https://github.com/SUSE/rmt/blob/master/engines/registration_sharing/package/rmt-server-trim-cache.service as well (as that's the only function that _actually_ expires those cache items).

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/221800/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 5c71441770c3c145e3d50aabbf0599c8c5b0d7c4 into e30273fcb614dc9ea2561a4d2a903122f5669f3b)

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+      response = http.request(scc_request)
+      unless response.code_type == Net::HTTPOK
+        logger.info "Could not get the system (#{login}) activations, error: #{response.message} #{response.code}"
+        response.message = SccProxy.parse_error(response.message) if response.message.include? 'json'
+        return { is_active: false, message: response.message }
+      end
+      scc_systems_activations = JSON.parse(response.body)
+      return { is_active: false, message: 'No activations.' } if scc_systems_activations.empty?
+
+      no_status_products_ids = scc_systems_activations.map {|act| act['service']['product']['id'] if (act['status'].nil? && act['expires_at'].nil?)}.flatten.compact
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:101]: Could you provide me some examples where you encountered this? That seems to be an issue in SCC data / API  if that happens.

**Review by digitaltom — APPROVED** [review]
Looks fine for me from reading the code + tests. 
Please give this a good real testing, because I wasn't able to run this at all, I only looked at the code. It would be really helpful to be able to have a local test setup, I'm sure we'll discover issues in the real deployment which could have been catched earlier with a local test. 

Additionally it would be useful to have a .changelog file for the publoud rpm package, so you could describe the changes there.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/221898/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 1cb96776901033bb790dd6f256f3182b24d7a06a into e30273fcb614dc9ea2561a4d2a903122f5669f3b)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/221921/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-run-tests. (Merged 00ea70d7fc8f41bdf51e77abf1a3e13c6604e40d into e30273fcb614dc9ea2561a4d2a903122f5669f3b)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/221923/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-run-tests. (Merged 00ea70d7fc8f41bdf51e77abf1a3e13c6604e40d into e30273fcb614dc9ea2561a4d2a903122f5669f3b)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/222099/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged ec604f0a7c86ac5b93fd672bcf90f1b97a572c4c into e30273fcb614dc9ea2561a4d2a903122f5669f3b)
